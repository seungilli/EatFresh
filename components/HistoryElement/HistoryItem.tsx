import { HistoryType } from "@/types/data";
import { View, StyleSheet } from "react-native";
import { Text } from "react-native-paper";

const styles = StyleSheet.create({
  historyItem: {
    flexDirection: "row",
    alignItems: "center",
    marginBottom: 10,
  },
  historyTime: {
    marginRight: 10,
  },
  historyText: {
    flex: 1,
    fontSize: 16,
    fontWeight: "bold",
  },
});

type HistoryItemProps = {
  item: HistoryType;
};

export default function HistoryItem({ item }: HistoryItemProps) {
  return (
    <View style={styles.historyItem}>
      <Text style={styles.historyTime}>{item.time}</Text>
      <Text style={styles.historyText}>{item.searchedText}</Text>
    </View>
  );
}
