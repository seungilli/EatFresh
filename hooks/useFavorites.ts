import { useState, useEffect, useCallback } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Meal } from '../types/data';

const useFavorites = () => {
    const [favorites, setFavorites] = useState<Meal[]>([]);

    const fetchFavorites = useCallback(async () => {
        try {
            const value = await AsyncStorage.getItem('favorite');
            const parsedFavorites = value != null ? JSON.parse(value) : [];
            setFavorites(Array.isArray(parsedFavorites) ? parsedFavorites : []);
        } catch (error) {
            console.error('Error fetching favorites:', error);
        }
    }, []);

    const addFavorite = useCallback(
        async (item: Meal) => {
            try {
                const existingFavorites = await AsyncStorage.getItem('favorite');
                let updatedFavorites = existingFavorites ? JSON.parse(existingFavorites) : [];
                if (!Array.isArray(updatedFavorites)) {
                    updatedFavorites = [];
                }
                updatedFavorites.push(item);
                await AsyncStorage.setItem('favorite', JSON.stringify(updatedFavorites));
                await fetchFavorites();
            } catch (error) {
                console.error('Error adding favorite:', error);
            }
        },
        [fetchFavorites]
    );

    const removeFavorite = useCallback(
        async (id: string) => {
            try {
                const existingFavorites = await AsyncStorage.getItem('favorite');
                let updatedFavorites = existingFavorites ? JSON.parse(existingFavorites) : [];
                if (Array.isArray(updatedFavorites)) {
                    updatedFavorites = updatedFavorites.filter((meal: Meal) => meal.idMeal !== id);
                    await AsyncStorage.setItem('favorite', JSON.stringify(updatedFavorites));
                    await fetchFavorites();
                }
            } catch (error) {
                console.error('Error removing favorite:', error);
            }
        },
        [fetchFavorites]
    );

    const clearFavorites = useCallback(async () => {
        try {
            await AsyncStorage.removeItem('favorite');
            await fetchFavorites();
        } catch (error) {
            console.error('Error clearing favorites:', error);
        }
    }, [fetchFavorites]);

    useEffect(() => {
        fetchFavorites();
    }, [fetchFavorites]);

    return {
        favorites,
        addFavorite,
        removeFavorite,
        clearFavorites,
        refreshFavorites: fetchFavorites,
    };
};

export default useFavorites;
