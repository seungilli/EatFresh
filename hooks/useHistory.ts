import { useState, useEffect, useCallback } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { HistoryType } from '@/types/data';

const useHistory = () => {
    const [history, setHistory] = useState<HistoryType[]>([]);

    const fetchHistory = useCallback(async () => {
        try {
            const value = await AsyncStorage.getItem('history');
            const parsedHistory = value != null ? JSON.parse(value) : [];
            setHistory(Array.isArray(parsedHistory) ? parsedHistory : []);
        } catch (error) {
            console.error('Error fetching history:', error);
        }
    }, []);

    const addHistoryItem = useCallback(
        async (item: HistoryType) => {
            try {
                const existingHistory = await AsyncStorage.getItem('history');
                let updatedHistory = existingHistory ? JSON.parse(existingHistory) : [];
                if (!Array.isArray(updatedHistory)) {
                    updatedHistory = [];
                }
                updatedHistory.push(item);
                await AsyncStorage.setItem('history', JSON.stringify(updatedHistory));
                await fetchHistory();
            } catch (error) {
                console.error('Error adding history:', error);
            }
        },
        [fetchHistory]
    );

    const clearHistory = useCallback(async () => {
        try {
            await AsyncStorage.removeItem('history');
            await fetchHistory();
        } catch (error) {
            console.error('Error clearing history:', error);
        }
    }, [fetchHistory]);

    useEffect(() => {
        fetchHistory();
    }, [fetchHistory]);

    return {
        history,
        addHistoryItem,
        clearHistory,
        refreshHistory: fetchHistory,
    };
};

export default useHistory;
