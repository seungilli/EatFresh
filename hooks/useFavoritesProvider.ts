import React, { createContext, useContext, ReactNode } from "react";
import useFavorites from "./useFavorites";

type FavoritesContextType = ReturnType<typeof useFavorites>;

const FavoritesContext = createContext<FavoritesContextType | undefined>(
    undefined
);

export const FavoritesProvider = ({ children }: { children: ReactNode }) => {
    const favorites = useFavorites();

    return React.createElement(
        FavoritesContext.Provider,
        { value: favorites },
        children
    );
};

export const useFavoritesContext = (): FavoritesContextType => {
    const context = useContext(FavoritesContext);
    if (!context) {
        throw new Error(
            "useFavoritesContext must be used within a FavoritesProvider"
        );
    }
    return context;
};
