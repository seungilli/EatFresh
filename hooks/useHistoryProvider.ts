import React, { createContext, useContext, ReactNode } from "react";
import useHistory from "./useHistory";

type HistoryContextType = ReturnType<typeof useHistory>;

const HistoryContext = createContext<HistoryContextType | undefined>(
    undefined
);

export const HistoryProvider = ({ children }: { children: ReactNode }) => {
    const history = useHistory();

    return React.createElement(
        HistoryContext.Provider,
        { value: history },
        children
    );
};

export const useHistoryContext = (): HistoryContextType => {
    const context = useContext(HistoryContext);
    if (!context) {
        throw new Error(
            "useHistoryContext must be used within a HistoryProvider"
        );
    }
    return context;
};
