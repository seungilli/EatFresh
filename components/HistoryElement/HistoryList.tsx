import { View, TouchableOpacity, StyleSheet, FlatList } from "react-native";
import { Avatar, Button, Card, Text } from "react-native-paper";
import HistoryItem from "./HistoryItem";
import { useHistoryContext } from "@/hooks/useHistoryProvider";

const styles = StyleSheet.create({
  sectionTitle: {
    fontSize: 18,
    fontWeight: "bold",
    marginVertical: 10,
  },
});

export default function HistoryList() {
  const { history, clearHistory } = useHistoryContext();

  return (
    <View>
      <Text style={styles.sectionTitle}>History</Text>
      <FlatList
        data={history}
        renderItem={({ item }) => <HistoryItem item={item} />}
        keyExtractor={(item) => item.time}
      />
    </View>
  );
}
