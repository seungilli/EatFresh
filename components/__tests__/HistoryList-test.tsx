import React from "react";
import renderer, { act } from "react-test-renderer";
import { FlatList } from "react-native";
import HistoryList from "../HistoryElement/HistoryList";
import HistoryItem from "../HistoryElement/HistoryItem";

jest.mock("@react-native-async-storage/async-storage", () => ({
  setItem: jest.fn(() => Promise.resolve()),
  getItem: jest.fn(() => Promise.resolve(JSON.stringify([]))),
  removeItem: jest.fn(() => Promise.resolve()),
  clear: jest.fn(() => Promise.resolve()),
  getAllKeys: jest.fn(() => Promise.resolve([])),
}));

jest.mock("@/hooks/useHistoryProvider", () => ({
  useHistoryContext: jest.fn(),
}));

const mockHistory = [
  { time: "2023-11-23T12:00:00Z", details: "History Item 1" },
  { time: "2023-11-24T14:00:00Z", details: "History Item 2" },
];

describe("<HistoryList />", () => {
  const mockUseHistoryContext =
    require("@/hooks/useHistoryProvider").useHistoryContext;

  beforeEach(() => {
    mockUseHistoryContext.mockReturnValue({ history: mockHistory });
  });

  it("should render", async () => {
    const tree = renderer.create(<HistoryList />);
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it("should render the correct number of HistoryItem components", async () => {
    const component = renderer.create(<HistoryList />);
    const root = component.root;

    const historyItems = root.findAllByType(HistoryItem);
    expect(historyItems.length).toBe(mockHistory.length);
  });

  it("should pass the correct data to HistoryItem components", async () => {
    const component = renderer.create(<HistoryList />);
    const root = component.root;

    const historyItems = root.findAllByType(HistoryItem);

    historyItems.forEach((itemComponent, index) => {
      expect(itemComponent.props.item).toEqual(mockHistory[index]);
    });
  });

  it("should render an empty list when there is no history", async () => {
    mockUseHistoryContext.mockReturnValue({ history: [] });

    const component = renderer.create(<HistoryList />);
    const root = component.root;

    const flatList = root.findByType(FlatList);
    expect(flatList.props.data.length).toBe(0);
  });
});
