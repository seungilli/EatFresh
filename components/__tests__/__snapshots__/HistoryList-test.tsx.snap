// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryList Component renders correctly with given history data 1`] = `
<RCTScrollView
  data={
    [
      {
        "details": "History Item 1",
        "time": "2023-11-23T12:00:00Z",
      },
      {
        "details": "History Item 2",
        "time": "2023-11-24T14:00:00Z",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 10,
          }
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "marginRight": 10,
              },
            ]
          }
        >
          2023-11-23T12:00:00Z
        </Text>
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "flex": 1,
                "fontSize": 16,
                "fontWeight": "bold",
              },
            ]
          }
        />
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 10,
          }
        }
      >
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "marginRight": 10,
              },
            ]
          }
        >
          2023-11-24T14:00:00Z
        </Text>
        <Text
          style={
            [
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "fontFamily": "System",
                "fontWeight": "400",
                "letterSpacing": 0,
              },
              {
                "writingDirection": "ltr",
              },
              {
                "flex": 1,
                "fontSize": 16,
                "fontWeight": "bold",
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
